### PR TITLE
bgp_address_family: Fix OSPF Redistribution Issues 

### PR DESCRIPTION
**Pull request by bentole**
##### SUMMARY

Fixes issues when redistributing OSPFv2 and OSPFv3 into BGP

- Fixed issue no commands generated
- Fixed `include-connected` option when redistributing OSPF in IPv6 AFI
- Fixed option and syntax for `E1` and `E2` routes
- Fixed option and syntax for `N1` and `N2` routes
- Fixed syntax and regex issues in OSPFv2 and OSPFv3 parsers
- Fixed idempotency issues
- Added appropriate negation of commands in overridden and replaced state

Related issue #780 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
ios_bgp_address_family

### CODE REVIEW

File: plugins/module_utils/network/ios/argspec/bgp_address_family/bgp_address_family.py [hunk 1]
```diff
@@ -668,14 +668,25 @@ class Bgp_address_familyArgs(object):  # pylint: disable=R0903
...
-                                                "type_1": {"type": "bool"},
-                                                "type_2": {"type": "bool"},
+                                                "external": {
+                                                    "type": "dict",
+                                                    "options": {
+                                                        "type_1": {"type": "bool"},
+                                                        "type_2": {"type": "bool"},
+                                                    },
+                                                },
+                                                "nssa_external": {
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/argspec/bgp_address_family/bgp_address_family.py:683]: @bentole we ideally deprecate the attributes which are not perfect by type and re-introduce a new set of attributes, to address it as a bugfix, or else they become a breaking change that requires a major release.

Comment (KB-perByte) [reply plugins/module_utils/network/ios/argspec/bgp_address_family/bgp_address_family.py:683]: You may refer to the PR https://github.com/ansible-collections/cisco.ios/pull/687/files for reference on how to deprecate, incase the tonne of things done in this PR confuse you let me know, I would push changes to this PR directly.
Regards

File: plugins/module_utils/network/ios/rm_templates/bgp_address_family.py [hunk 2]
```diff
@@ -2514,41 +2531,53 @@ def __init__(self, lines=None, module=None):
...
-                (\s(?P<type_1>1))?
-                (\s(?P<type_2>2))?
-                (\s(?P<external>external))?
-                (\s(?P<internal>internal))?
-                (\s(?P<nssa_external>nssa-external))?
+                \s+redistribute\sospfv3\s(?P<process_id>\S+)
                 (\smetric\s(?P<metric>\d+))?
+                (\smatch)?
+                (\s(?P<internal>internal))?
+                (\s(?P<ext_type_1>external\s1))?
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/rm_templates/bgp_address_family.py:2538]: As said we would add fresh attributes with correct options in argspec, we would never show the deprecated attributes in facts but rather handle the option flowing in from the playbook in the config side code, and just render back the correct attributes back in facts.

**Review by KB-perByte — CHANGES_REQUESTED** [review]

File: plugins/module_utils/network/ios/argspec/bgp_address_family/bgp_address_family.py (hunk 1 above)
Comment (bentole) [reply plugins/module_utils/network/ios/argspec/bgp_address_family/bgp_address_family.py]: Hi @KB-perByte ! I have made an attempt, emphasis on attempt, to address the issues you were requesting. Ideally I'd like to keep the `external` and `nssa_external` params just changing their type from `bool` to `dict`, but that's probably not possible until the old ones are removed?  Instead I introduced `externals` and `nssa_externals`. Also made an attempt to rewrite the changelog fragment. Hope I'm closer to compliance 😃

**Comment by KB-perByte** [comment]
I am holding this PR in this release. Sorry for the inconvenience!

**Comment by bentole** [comment]
Hi @KB-perByte ! I need to be spoon-fed 😅 If I understand correctly stuff should be re-arranged to look something like this:

```
redistribute:
  - connected:
      metric: 10
      set: true
  - ospfs:
      - process_id: 123
      - process_id: 124
        route_map: bar
```
?

I was thinking since redistribute is already a list of dicts it's perfectly fine to repeat elements. This play works perfectly fine with all states:

```
redistribute:
  - connected:
      metric: 10
      set: true
  - ospf:
      process_id: 123
  - ospf:
      process_id: 124
      route_map: bar
```

If the latter is not consistent with assertions I should probably just rewrite the integration test to redistribute the two processes in separate tasks? If we are to introduce a parent attribute like `ospfs` we should probably do the same with isis and eigrp and then we are doing/fixing something currently not addressed by this PR. Probably better to just rewrite the test. What say ye, boss? :)

**Comment by KB-perByte** [comment]
@bentole sorry for the delayed reply! 
You are right, it works and the states work fine. On second thought let's get this merged as of now. And I'll log it somewhere that we need to redo redistribution, it has been introduced as a list waaaay back, but it should have been a dict that has list attributes when needed. I remember adding a [turn](https://github.com/ansible-collections/cisco.ios/blob/b9532a5807c6e1d49e06e1d0b46091e80a3240d4/plugins/module_utils/network/ios/config/bgp_address_family/bgp_address_family.py#L345) specifically for having this work.
it should be as simple as as top-level [dict](https://github.com/ansible-collections/cisco.ios/blob/b9532a5807c6e1d49e06e1d0b46091e80a3240d4/plugins/module_utils/network/ios/argspec/bgp_address_family/bgp_address_family.py#L591) and ospf as [list](https://github.com/ansible-collections/cisco.ios/blob/b9532a5807c6e1d49e06e1d0b46091e80a3240d4/plugins/module_utils/network/ios/argspec/bgp_address_family/bgp_address_family.py#L668) with elements as dict. Basically if we make an attribute list and fix all the possible entries it is somewhat pointless (wrong) from the model design standpoint! 
Sorry, I made you confused.
Getting this PR merged.
Thanks! :)

**Review by KB-perByte — APPROVED** [review]

**Pull request merged**